### PR TITLE
feat(editor): Add HTTP request nodes for credentials without a node

### PR DESCRIPTION
**Pull request by elsmr**
Github issue / Community forum post (link here to close automatically):

### CODE REVIEW

**Review by mutdmour — CHANGES_REQUESTED** [review]
Thanks for your PR. Can you please make sure to add tests to cover this new feature? Unit tests? Component tests? E2E tests?

File: packages/editor-ui/src/components/ParameterInputList.vue
Comment (mutdmour) [inline packages/editor-ui/src/components/ParameterInputList.vue:29]: would not this now show for old versions of http node? I imagine we need something like
`parameter.type === 'curlImport' && ((nodeTypeName === 'n8n-nodes-base.httpRequest' &&
					nodeTypeVersion >= 3) || isHTTPRequestedBase(nodeType))`

File: packages/core/bin/generate-ui-types
Comment (mutdmour) [inline packages/core/bin/generate-ui-types:39]: do we really need to clone the entire node properties? can we specify it inherits a specific node and version and update the n8n to support that instead of returning 20 different versions of the same node? 
as this could potentially explode in number in the future.. for example if we release a new version of http node and want to use that here, the number of imported nodes would double

File: packages/editor-ui/src/constants.ts
Comment (mutdmour) [inline packages/editor-ui/src/constants.ts:175]: this has one function.. to not show description.. is it used for anything else? seems maybe it's unneeded and we should do that some other way.

File: packages/core/bin/generate-ui-types
Comment (mutdmour) [reply packages/core/bin/generate-ui-types:39]: @netroy would love your feedback on this.. since you have given a lot of thought to n8n memory problems..

**Review by mutdmour — COMMENTED** [review]
Love this feature. Thanks for your work on this. Please make sure to add tests..

File: packages/core/bin/generate-ui-types
Comment (elsmr) [reply packages/core/bin/generate-ui-types]: It is being shown in the UI, tweaked the frontend logic to make it more clear in my latest commit

File: packages/editor-ui/src/components/ParameterInputList.vue
Comment (elsmr) [reply packages/editor-ui/src/components/ParameterInputList.vue:29]: Only the HTTP Request V3 Node has a parameter with type == `curlImport`, so old versions don't show it. I thought these checks were not needed, as no other node has `curlImport` as a parameter type. Let me know if I'm missing something, I'll revert it

File: packages/editor-ui/src/constants.ts
Comment (elsmr) [reply packages/editor-ui/src/constants.ts]: Removed it

File: packages/core/bin/generate-ui-types
Comment (elsmr) [reply packages/core/bin/generate-ui-types]: This does indeed add a number of new nodes to `types/nodes.json` (though I also fixed a bug to not include duplicate nodes 😄). I'm currently trying to see if I can add some kind of `extend` feature to Nodes like we have for credentials.

File: packages/editor-ui/src/components/ParameterInputList.vue
Comment (mutdmour) [reply packages/editor-ui/src/components/ParameterInputList.vue:29]: Perfect. much better this way

File: packages/core/bin/generate-ui-types
Comment (netroy) [reply packages/core/bin/generate-ui-types]: This does however does change the content of these files, so I still need to validate that those changes don't break any existing functionality.

Comment (mutdmour) [reply packages/core/bin/generate-ui-types]: @netroy would not this also be pulling these nodes during execution?

Comment (mutdmour) [reply packages/core/bin/generate-ui-types]: @elsmr if @netroy does not have a problem with this here, feel free to ignore my feedback.. hopefully you have not wasted a lot of time on this..

Comment (netroy) [reply packages/core/bin/generate-ui-types]: just FYI: I still haven't gotten around to reviewing this yet, my earlier comment is only about the memory-usage concerns.

Comment (elsmr) [reply packages/core/bin/generate-ui-types]: Just in time, I was about to have another look at it today, but I'll hold off until you review the changes to `packages/nodes-base/dist/types` @netroy

**Comment by netroy** [comment]
The code changes look good, but the frontend doesn't seems to work for me for some reason. 
Investigating now. if it's my local setup that's broken, then I can approve this PR.

File: packages/cli/src/CredentialsHelper.ts
Comment (netroy) [inline packages/cli/src/CredentialsHelper.ts:7]: Can we please move all the `import/order` auto-fixes into a separate PR? that not only makes reviewing this easier, but also helps with bisecting for issues in the future.

File: packages/core/src/DirectoryLoader.ts
Comment (netroy) [inline packages/core/src/DirectoryLoader.ts:185]: Since this assumes that nodes will always be loaded before credentials, I think we should add a test for it.

File: packages/core/src/DirectoryLoader.ts
Comment (netroy) [inline packages/core/src/DirectoryLoader.ts:347]: since we swapped the order here, we should check that we aren't breaking any of the community nodes.

**Review by netroy — COMMENTED** [review]
Left some comments. Still trying to figure out why the editor is broken for me only on this branch.

File: packages/nodes-base/nodes/HttpRequest/HttpRequest.node.ts
Comment (netroy) [inline packages/nodes-base/nodes/HttpRequest/HttpRequest.node.ts:13]: can we also please move this icon change into a separate PR 🙏🏽

File: packages/cli/src/CredentialsHelper.ts
Comment (elsmr) [reply packages/cli/src/CredentialsHelper.ts]: Removed from this PR

**Review by netroy — DISMISSED** [review]
The backend bits look good to me.
I'll add unit tests for `DirectoryLoader` in another PR.

**Review by netroy — DISMISSED** [review]

**Comment by netroy** [comment]
before we merge this, we should run the test-workflows CI, and update the PR title

File: packages/editor-ui/src/mixins/workflowRun.ts
Comment (mutdmour) [inline packages/editor-ui/src/mixins/workflowRun.ts:233]: not enough to do this on the FE.. the BE needs to support this new type.. 
the workflow cannot be activated
Screenshot 2023-11-08 at 13 44 59

also shows as http node in execution.. (which might be forgivable but above is not)
Screenshot 2023-11-08 at 13 46 07

I would say it's simpler to keep the type as `n8n-nodes-base.httpRequest` and add another property 
Screenshot 2023-11-08 at 13 55 19

File: packages/nodes-base/nodes/HttpRequest/V3/HttpRequestV3.node.ts
Comment (mutdmour) [inline packages/nodes-base/nodes/HttpRequest/V3/HttpRequestV3.node.ts:36]: why was this needed?

File: packages/editor-ui/src/utils/credentialOnlyNodes.ts
Comment (mutdmour) [inline packages/editor-ui/src/utils/credentialOnlyNodes.ts:56]: i18n - would be great to add this to en.json

File: packages/editor-ui/src/components/Node/NodeCreator/composables/useActionsGeneration.ts
Comment (mutdmour) [inline packages/editor-ui/src/components/Node/NodeCreator/composables/useActionsGeneration.ts:292]: console log

**Review by mutdmour — CHANGES_REQUESTED** [review]
Needs some more work.

**Comment by netroy** [comment]
Once this is merged, I think we should add a workflow to the [tests repo](https://github.com/n8n-io/test-workflows), to test some of these new virtual node types.

File: packages/editor-ui/src/mixins/workflowRun.ts
Comment (elsmr) [reply packages/editor-ui/src/mixins/workflowRun.ts]: I think the second one is because the iframe points to localhost:5678 (still running old frontend unless you rebuild)

Comment (mutdmour) [reply packages/editor-ui/src/mixins/workflowRun.ts]: no, I was testing on 5678 not 8080.. as you can see in the screenshot, the type is `httpRequest` in the execution, and that's why that icon is showing.

**Comment by elsmr** [comment]
> Once this is merged, I think we should add a workflow to the [tests repo](https://github.com/n8n-io/test-workflows), to test some of these new virtual node types.

I'll pick that up 😄

File: packages/editor-ui/src/constants.ts
Comment (mutdmour) [inline packages/editor-ui/src/constants.ts:160]: can we just use the latest version when the http node is created?

Comment (elsmr) [reply packages/editor-ui/src/constants.ts:160]: No, because the code in `utils/credentialOnlyNodes.ts` modifies the http node properties available in v4.1. If a new version of the http request node gets released, and has different properties, that code will break.

Instead the code will need to be updated when a new version of the http request node is released (but if it's not, nothing will break, credential nodes will keep working on 4.1)

File: packages/editor-ui/src/utils/credentialOnlyNodes.ts
Comment (elsmr) [reply packages/editor-ui/src/utils/credentialOnlyNodes.ts:6]: Yes, the frontend still handles it as a special type, only when saving/loading the workflow from the backend it's transformed to `extendsCredential` and `type: 'n8n-nodes-base.httpRequest'`.

My reasoning was that `nodeTypesStore.getNodeTypeByName` only takes a `nodeTypeName: string`. If I were to use extendsCredential, I would need to add an argument and edit all callsites (about 60) to send that new argument. Same for the NodeCreator, it only sends nodeTypeName when a node is added, I would also need to change a lot of code to include extendsCredential.

**Review by mutdmour — APPROVED** [review]
LGTM. Thank you for addressing all my feedback. Awesome job

**Comment by janober** [comment]
Got released with [`n8n@1.17.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.17.0)

**Pull request merged**
